### PR TITLE
Composing text shouldn't be part of undo/redo

### LINKED ISSUE #106473: Text field undo incorrectly cycles through all composing variants
## Steps to Reproduce

1. Execute `flutter run` on the code sample
2. Set Japanese IME (this is reproduced on Windows but it should be the same on other OSs)
3. Type some text (e.g. "anatano")
4. Use key arrows up/down to go through the different IME options, select one
5. Press Ctrl+Z to undo

**Expected results:** A single undo with Ctrl+Z goes back to empty text i.e. during composing, no undo states are stored.

**Actual results:** Each Ctrl+Z press goes to the previously traversed IME option.

undo-japanese-ime

<details>
<summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(primarySwatch: Colors.blue),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text(widget.title)),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: const <Widget>[
            Text('Type stuff below'),
            Expanded(
              child: TextField(maxLines: null),
            ),
          ],
        ),
      ),
    );
  }
}
```

</details>

<details>
  <summary>Logs</summary>

```
> flutter doctor -v
[√] Flutter (Channel master, 3.1.0-0.0.pre.1354, on Microsoft Windows [Version 10.0.19044.1708], locale en-US)
    • Flutter version 3.1.0-0.0.pre.1354 on channel master at C:\flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision a30012b275 (9 hours ago), 2022-06-22 17:04:07 -0700
    • Engine revision fc08bf45b0
    • Dart version 2.18.0 (build 2.18.0-216.0.dev)
    • DevTools version 2.14.0

[X] Android toolchain - develop for Android devices
    X Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.11.15)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.11.32510.428
    • Windows 10 SDK version 10.0.19041.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[√] VS Code, 64-bit edition (version 1.68.1)
    • VS Code at C:\Program Files\Microsoft VS Code
    • Flutter extension version 3.42.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19044.1708]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 102.0.5005.115
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 101.0.1210.39

[√] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 2 categories.
```

</details>

### PR DESCRIPTION
**Pull request by justinmc**
When using a composing IME, composing text should not be counted as a part of the undo/redo stack.

| Before (from issue) | After |
| --- | --- |
|  | out |

CC @tgucio 

Fixes https://github.com/flutter/flutter/issues/106473

### CODE REVIEW

**Review by LongCatIsLooong — COMMENTED** [review]
> When using a composing IME, composing text should not be counted as a part of the undo/redo stack.

That depends on how the IME uses the composing region? Gboard puts everything in a composing region so this would prevent undo-redo from working if the user is using gboard.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -4693,6 +4693,11 @@ class _TextEditingHistoryState extends State<_TextEditingHistory> {
       return;
     }
 
+    // Composing text is not counted in history coalescing.
+    if (widget.controller.value.composing != null && !widget.controller.value.composing.isCollapsed) {
```
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/widgets/editable_text.dart:4697]: Good catch, I feel like that should be caught by the analyzer...

**Comment by justinmc** [comment]
@LongCatIsLooong Shucks good point.

It looks like Gboard does not have undo/redo capabilities [0].  I tried hooking up a keyboard to my Android phone and ctrl-z does indeed work on both native and Flutter.  If I go so far as to type a bunch of text with Gboard, then plug in a keyboard and start pressing ctrl-z, Flutter's coalescing happens only between words, while native sometimes undoes only part of a word.  That's incorrect but a pretty subtle and niche difference.

Do you think it's better if I check the platform and allow coalescing during composing on mobile, or another approach?  Or is the current way best?

[0] The Google Keep app has its own undo/redo buttons, and it does coalesce partially composed words. A Flutter dev could implement that if they wanted, though.  Other apps have no way to undo/redo with Gboard.

**Comment by justinmc** [comment]
@LongCatIsLooong I updated this PR to allow coalescing of composing text on Android only.  This isn't perfect but it seems to handle the more common use case:

| Scenario | This PR as-is, which doesn't coalesce composing text | Only coalesce composing text on Android, not elsewhere |
| --- | --- | --- |
| Typing English on Android on a hardware keyboard with Gboard enabled | ❌ Coalesces whole words. | ✅ Coalesces partial words |
| Typing CJK on Android on a hardware keyboard with Gboard enabled | ✅ Coalesces final characters. | ✅ Coalesces final characters. |
| Typing CJK on Android on a hardware keyboard with a software keyboard that shows composing text in the field | ✅ Coalesces final characters. | ❌ Coalesces composing text. |
| Typing English everywhere else | ✅ Coalesces partial words. | ✅ Coalesces partial words. |
| Typing CJK everywhere else | ✅ Coalesces final characters. | ✅ Coalesces final characters. |

Let me know what you think.

**Review by LongCatIsLooong — APPROVED** [review]

**Pull request merged**